sql/job_stat_history_log_retention.sql

```diff
@@ -4,23 +4,97 @@
 
 -- A retention policy is set up for the table _timescaledb_internal.job_errors (Error Log Retention Policy [2])
 -- By default, it will run once a month and and drop rows older than a month.
+
+CREATE OR REPLACE FUNCTION _timescaledb_functions.job_history_bsearch(search_point TIMESTAMPTZ) RETURNS BIGINT
+AS
+$$
+DECLARE
+  id_lower BIGINT;
+  id_upper BIGINT;
+  id_middle BIGINT DEFAULT 0;
+  id_middle_next BIGINT;
+  range_start TIMESTAMPTZ;
+  range_end TIMESTAMPTZ;
+  target_tz TIMESTAMPTZ;
+BEGIN
+
+  SELECT COALESCE(min(id), 0), COALESCE(max(id), 0)
+  INTO id_lower, id_upper
+  FROM _timescaledb_internal.bgw_job_stat_history;
+
+  IF id_lower = 0 AND id_upper = 0 THEN
+    RETURN NULL;
+  END IF;
+
+  -- We want the first entry in the table where execution_finish is >= search_point
+  WHILE id_lower < id_upper LOOP
+    id_middle := id_lower + (id_upper - id_lower) / 2;
+
+    SELECT execution_finish
+    INTO target_tz
+    FROM _timescaledb_internal.bgw_job_stat_history
+    WHERE id = id_middle;
+
+    -- If the id_middle is not found, shift to the next id
+    IF NOT FOUND THEN
+      SELECT execution_finish, id
+      INTO target_tz, id_middle_next
+      FROM _timescaledb_internal.bgw_job_stat_history
+      WHERE id >= id_middle AND id <= id_upper
+      ORDER BY id LIMIT 1;
+
+      IF NOT FOUND THEN
+        RETURN NULL;
+      END IF;
+
+      id_middle := id_middle_next;
+    END IF;
+
+    IF target_tz >= search_point THEN
+      id_upper := id_middle;
+    ELSE
+      id_lower := id_middle + 1;
+    END IF;
+  END LOOP;
+
+  RETURN id_lower;
+END
+$$
+LANGUAGE plpgsql SET search_path TO pg_catalog, pg_temp;
+
 CREATE OR REPLACE FUNCTION _timescaledb_functions.policy_job_stat_history_retention(job_id integer, config JSONB) RETURNS integer
 LANGUAGE PLPGSQL AS
 $BODY$
 DECLARE
-    drop_after INTERVAL;
     numrows INTEGER;
+    search_point TIMESTAMPTZ;
+    id_found BIGINT;
 BEGIN
-    drop_after := config->>'drop_after';
+  search_point := now() - (config->>'drop_after')::interval;
 
-    DELETE
-    FROM _timescaledb_internal.bgw_job_stat_history
-    WHERE execution_finish < (now() - drop_after);
+  -- We need to prevent concurrent changes on this table when running this retention job
+  LOCK TABLE _timescaledb_internal.bgw_job_stat_history IN SHARE ROW EXCLUSIVE MODE;
```

Comment:
```suggestion
  LOCK TABLE _timescaledb_internal.bgw_job_stat_history IN EXCLUSIVE MODE;
```
I've made a mistake the correct explicit lock here is `EXCLUSIVE MODE`: https://www.postgresql.org/docs/current/explicit-locking.html

Reply:
Just wondering if we should fallback to a `DELETE` if for some reason the lock is not taken. We can set a local lock timeout (200ms or 300ms) and if the lock is not taken then we fallback to a `DELETE` instead. We keep use the `job_history_bsearch` because with it we can use the primary key index to filter rows (the root cause of this implementation since the current `DELETE` filter the rows based on the `execution_finish` columns that is not indexed).

Reply:
I changed it to an AccessExclusiveLock and added a lock timeout.